### PR TITLE
Threshold Activation of Protocol Improvements (TAPI)

### LINKED ISSUE #38: WIP covering a mechanism for automatic activation of consensus rules upon generalized support signal
A briefly described here: https://github.com/witnet/witnet-rust/pull/1829#pullrequestreview-572212003

Witnet-rust issue for tracking reference implementation of this: https://github.com/witnet/witnet-rust/issues/1805

### PR DESCRIPTION
**Pull request by aesedepece**
Close #38

### CODE REVIEW

**Comment by lrubiorod** [comment]
I assign you the wip number WIP-0014

**Comment by aesedepece** [comment]
I'll fixing the pseudocode examples, as our friend Dr. CPU noticed they are wrong in many ways :rofl:

**Comment by aesedepece** [comment]
- Changed the default signaling bitmask to `0b00000000000000000000000000000001` for telling apart implementers of this feature when activating it.
- Changed the pseudocode snippets to JavaScript and verified that they actually run.
- This is done for now on my side. I hereby request merging, and then I'll be working on an adoption plan and request promotion to "Proposed".

**Comment by aesedepece** [comment]
Last commit:
- introduces `first_signaling_epoch`
- explicitly specifies which is the first block with new rules after an improvement has been activated

File: wip-0014.md [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+is almost impossible to reach.
+
+A clear example of an unintended network halt during a coordinated upgrade is the  activation of [WIP-0009], [WIP-0011]
+and [WIP-0012] on 28 April 2021. Despite these WIPs receiving overall positive comments and no major objections from
+the community, by the time they were expected to enter into force, only some 45% of the network had upgraded to the
+`1.2.x` version of [witnet-rust]. This caused the network to split in two, with both sides of the split being unable
+to achieve the 2/3rds consensus required for resuming block consolidation. Only after 50 hours the percentage of
+upgraded nodes had grown closer to 2/3rds, a superblock was eventually consolidated by the supermajority chain, a
+fresh new *Active Reputation Set (ARS)* emerged, and consensus continued as usual, with over 90% average convergence
+and no single chain rollback ever since.
```
Comment (lrubiorod) [inline wip-0014.md:48]: Not anymore 😢

**Review by lrubiorod — APPROVED** [review]

**Pull request merged**